### DISCUSSION TITLE
Platform Support - Real world

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#14188] [General]
Hi folks,

**Context:**
I have an application written in WPF. It is an ERP application that primarily runs on Windows, with some instances on MacOS powered by Parallels.

I have plans to rewrite the entire UI in Avalonia. The idea is to recreate the same screens from scratch, migrating the entire application from WPF to Avalonia.

My goal is to run this on desktop platforms (Windows, Linux, and MacOS) as well as on mobile devices (specifically larger screens such as iPads and Android tablets). It would be great to have this available on WebAssembly (Wasm).

**Question:**
How feasible is this? How "business-ready" or "perhaps in the future" are each platform? 

I understand that the desktop transition is well-established, but what about iOS and Android (again, focusing on larger screens like tablets/iPads) and WebAssembly (Wasm)?

---

In order to make it less hard to think, here's some informations of my screens:

I open all my window inside an TabItem.

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/bd09c7c0-3843-4cc1-867f-efa3a4f0d27f)

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/c2a3c2d1-de89-48d1-9cc4-1efcefaa1c32)

For searches inside an instead of Show() or ShowDialog() on my Window I put the content inside an SimpleChildWindow (MahApps).

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/e3eb3e88-e68b-47c8-af45-312527997401)

**Comment by thevortexcloud** [comment]
Desktop is fairly stable at this point and should be fairly straightforward to port. However all the other platforms you mentioned require a SPA model to present your content (since they have no concept of windows). Which will likely require you to re architect things. Mobile platforms are also usable, but they are also fairly barebones out of the box. EG things like hamburger menus are not included by default. So you need to either make things like that yourself or use third party libraries. EG this one is meant to provide most of the features people expect on mobile: https://github.com/AvaloniaInside/Shell

You also likely need some sort of routing system for SPA. It's not too hard to write your own or use an existing one like what is found in ReactiveUI. I personally write my own since it's easier to make it behave how I want.

Web assembly is also pretty much the same.

On a side note, if you are willing to pay for it: Avalonia also has XPF which is a port of WPF to Avalonia's rendering and input system. Allowing you to run your WPF app without changing much/any code on other platforms.

**Comment by marcosgerene** [comment]
@thevortexcloud  Thanks for your time!

> Desktop is fairly stable at this point and should be fairly straightforward to port

I was pretty sure of this.

>  However all the other platforms you mentioned require a SPA model to present your content (since they have no concept of windows).

As I said, my application runs on a single window. Every new window is treated as a UserControl that is opened inside this main windows in a TabItem or in a ChildWindow (https://github.com/punker76/MahApps.Metro.SimpleChildWindow).

Does it solve the SPA problem?

Once it is solved, is Avalonia mature enough to mobile (big screens) and WASM? How ready to business is it?

Edit:
>In a side note, if you are willing to pay for it: Avalonia also has XPF which is a port of WPF to Avalonia's rendering and input system.

XPF is out of question by now. Rewrite in Avalonia is cheaper as long term.

Thank you again.

**Reply by thevortexcloud** [reply]
> Does it solve the SPA problem?

It does actually. It's probably not particularly great UX for a mobile device, but it will probably work until you can come up with something better/tweak it. Avalonia does have the [`OnPlatform` and `OnFormFactor`](https://docs.avaloniaui.net/docs/guides/platforms/platform-specific-code/xaml) markup extension which you can probably use to switch things around. In any case, that library you linked seems to be WPF only so you will probably need to either port that or make your own system. As a starting point, [the Dialog Hos](https://github.com/AvaloniaUtils/DialogHost.Avalonia)t works pretty well in my experience.

> Once it is solved, is Avalonia mature enough to mobile (big screens) 

I can't say I have made anything particularly complicated with it on mobile. But it all seemed to work when I last tried. Most third party controls however are designed for desktop usage which can make them behave weird/not great when used with a touch screen. Avalonia also relies on using Xamarin APIs and Xamarin/Maui essentials for a lot of mobile specific APIs. Both of which are made by Microsoft.

> WASM

Web Assembly requires .NET 8 _or_ some manual tweaking of some files to work properly with .NET 7. But from what I have heard it also works pretty well once you get it up and running. I can't say I have messed around with it too much myself though. It is however officially stable according to Avalonia. 

That said: Web assembly is also fairly sandboxed due to it being a web app which tends to throw people off. As a result it also has a lot of limitations compared to the desktop .NET runtime, which may require rewriting and abstracting a lot of code. There is a somewhat brief list [here](https://github.com/dotnet/runtime/blob/main/src/mono/wasm/features.md) made by Microsoft of some of the limitations with Web Assembly. The biggest one I imagine you may run into with an ERP app is database communication. There is no way to directly talk to a database in web assembly (unless you use SQLite) that I know of. As such you will need to proxy everything via some sort of networked API. Although if your app was designed sometime in the last ~15 years it probably does that already. 

Also compiling your app for AOT with trimming may also result in breaking a lot of your code. But you kind of need to do that to have somewhat reasonable boot times with web assembly. It's best to avoid using reflection (and libraries that use reflection) to reduce the chance of that happening.

**Comment by marcosgerene** [comment]
> It does actually. It's probably not particularly great UX for a mobile device, but it will probably work until you can come up with something better/tweak it.

Are you talking about performance?

> As a starting point, the Dialog Host works pretty well in my experience.

I did know the component. In my studies I will use Avalonia, Avalonia.DataGrid, Avalonia.Material and DialogHost.

>  Most third party controls however are designed for desktop usage which can make them behave weird/not great when used with a touch screen.

I will use Avalonia.DataGrid, Avalonia.Material and DialogHost, I don't know  enough about them on Mobile/Wasm. DataGrid is my Achilles' heel (I think  so). That is why I am targeting bigger screens.

>WASM

My application runs on .NET 8. We use PostgreSQL via Npgsql. It is just an client for a server database, I would guess it works without any problems, I will check this on Npgsql project.

**Reply by thevortexcloud** [reply]
> Are you talking about performance?

That's less of a UX problem and more of a functional one. But yes that may also be a problem. A lot of mobile devices are significantly slower than desktops. I was thinkng about how the actual experience of using a bunch of nested windows like what you have on a mobile device. I have seen other ERPs do a similar thing and it's fairly annoying to navigate, even on large screens. 

>  We use PostgreSQL via Npgsql. It is just an client for a server database

It's an ADO client. Which means it _likely_ (I have not used that specific ADO client) opens a direct socket connection to the database server. This wont work in web assembly as you can't open raw sockets. You will probably need to rewrite all your database calls to be sent via a web socket or HTTP request to some proxy server that can then call the database for you (you should probably do this anyway since opening up direct access to a database to the internet is generally a bad idea). Spinning up a simple ASP Core server with a REST API is probably the easiest route (make sure you have some sort of authentication though, EG JWT). As a bonus of doing that, your ERP would also get a user callable API service. Which is considered by a lot of companies to be an essential feature for an ERP (since it lets them do bulk imports and exports of data from what ever weird third party system they use). But making it usable for that sort of thing would mean moving most of your logic out of the client and into the server.

**Comment by marcosgerene** [comment]
>UX

I will have and do some adjustments. Once it will be build from scratch (UI/UX) it is easier to test in all platforms and make it as smooth as possible for an ERP.

> WASM

It is better to set aside for now.
